index.js

```diff
@@ -70,20 +71,111 @@ function format(text, opts, addAlignmentSize) {
+    case "ForStatement":
+    case "IfStatement":
+    case "LabeledStatement":
+    case "ReturnStatement":
+    case "SwitchStatement":
+    case "ThrowStatement":
+    case "TryStatement":
+    case "VariableDeclaration":
+    case "WhileStatement":
+    case "WithStatement":
+      return true;
+  }
+  return false;
+}
+
+function calculateRange(text, opts, ast) {
+  // Contract the range so that it has non-whitespace characters at its endpoints.
+  // This ensures we can format a range that doesn't end on a node.
+  const rangeStringOrig = text.slice(opts.rangeStart, opts.rangeEnd);
+  const rangeStringOrigReversed = rangeStringOrig.split("").reverse().join("");
+  const startNonWhitespace = Math.max(
+    opts.rangeStart + rangeStringOrig.search(/\S/),
+    opts.rangeStart
+  );
+  const endNonWhitespace = Math.min(
+    opts.rangeEnd - rangeStringOrigReversed.search(/\S/),
+    opts.rangeEnd
+  );
+
+  const startNode = findNodeByOffset(ast, startNonWhitespace);
+  const endNode = findNodeByOffset(ast, endNonWhitespace);
+  const rangeStart = Math.min(util.locStart(startNode), util.locStart(endNode));
+  const rangeEnd = Math.max(util.locEnd(startNode), util.locEnd(endNode));
+
+  const rangeString = text.slice(rangeStart, rangeEnd);
+
+  // Try to extend the range backwards to the beginning of the line.
+  // This is so we can detect indentation correctly and restore it.
   // Use `Math.min` since `lastIndexOf` returns 0 when `rangeStart` is 0
-  const rangeStart = Math.min(
-    opts.rangeStart,
-    text.lastIndexOf("\n", opts.rangeStart) + 1
+  const rangeStart2 = Math.min(
+    rangeStart,
+    text.lastIndexOf("\n", rangeStart) + 1
   );
-  // Use `text.length - 1` as the maximum since `indexOf` returns -1 if `fromIndex >= text.length`
-  const fromIndex = Math.min(opts.rangeEnd, text.length - 1);
-  const nextNewLineIndex = text.indexOf("\n", fromIndex);
-  const rangeEnd = (nextNewLineIndex < 0 ? fromIndex : nextNewLineIndex) + 1; // Add one to make rangeEnd exclusive
-
-  if (0 < rangeStart || rangeEnd < text.length) {
-    const rangeString = text.substring(rangeStart, rangeEnd);
-    const alignmentSize = getAlignmentSize(
+  const indentString = text.slice(rangeStart2, rangeStart);
+  try {
+    if (indentString.trim() == "") {
+      return {
+        rangeStart: rangeStart2,
+        rangeEnd: rangeEnd,
+        rangeString: text.slice(rangeStart2, rangeEnd)
+      };
+    }
+  } catch (err) {
+    // do nothing
+  }
+
+  return {
+    rangeStart: rangeStart,
+    rangeEnd: rangeEnd,
+    rangeString: rangeString
+  };
+}
+
+function formatRange(text, opts, ast) {
+  if (0 < opts.rangeStart || opts.rangeEnd < text.length) {
+    const range = calculateRange(text, opts, ast);
+    const rangeStart = range.rangeStart;
+    const rangeEnd = range.rangeEnd;
+    const rangeString = range.rangeString;
+
+    const alignmentSize = util.getAlignmentSize(
       rangeString.slice(0, rangeString.search(/[^ \t]/)),
```

Comment:
You are doing work above with rangeStart2 and then re-doing it again here. What about computing the indentString above and passing it down here?